### PR TITLE
Misc Visualizer fixes and enhancements

### LINKED ISSUE #1086: Visualizer: Misc enhancements and fixes
### Description of the task

Misc fixes and enhancements to the Visualizer

### Acceptance criteria

- [x] Fix visual bug when the Visualizer first starts up. The AI tab looks funny
- [x] PlayInfo shouldn't be editable
- [x] Zooming now zooms to the mouse rather than the middle of the window
- [x] Focus on the field during startup
- [x] Give the Visualizer a constant update rate
- [x] Visualizer sometimes segfaults on shutdown or locks up and stops full_system from shutting down. Fix this

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

*Enancements*
* Zooming now zooms on the mouse rather than the center of the view
* Properly focus on the field during startup

*Fixes*
* Fixed visual bug that caused the `Parameters` tab to show through the `AI Control` tab on startup
* Made the `PlayInfo` text field not editable
* Gave the Visualizer a constant update rate
* Fixed issues that caused the Visualizer to sometimes segfault on destruction, due to destroying components out of order

Overall the `VisualizerWrapper` _should_ be simpler now, since communication with the `Visualizer` instance is done using `ThreadSafeBuffers`, so lots of pointers, promises, and extra stuff was able to be removed or contained purely in the `run_visualizer` thread.

I think this is a better architecture so will probably document the new structure in #1087

### Testing Done

Manual verification

### Resolved Issues

resolves #1086 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
Tagging more people than usual since I don't know exactly who's free for reviews right now.

File: src/software/visualizer/widgets/visualizer.cpp [hunk 1]
```diff
@@ -1,27 +1,79 @@
...
 }
+
+void Visualizer::updateDrawViewArea()
+{
+    std::optional<Rectangle> view_area = view_area_buffer->popLeastRecentlyAddedValue();
+    if (view_area)
+    {
+        main_widget->setDrawViewArea(createQRectF(view_area.value()));
+    }
+}
```
Comment (jsn0716) [inline src/software/visualizer/widgets/visualizer.cpp]: Nit: no new line at the end of file

**Review by jsn0716 — CHANGES_REQUESTED** [review]
Just a couple nits, otherwise looks good to me.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Tried it out locally and worked well for me, just a few small comments.

File: src/software/visualizer/widgets/visualizer.cpp [hunk 2]
```diff
@@ -1,27 +1,79 @@
...
+      view_area_buffer(view_area_buffer),
+      update_timer_interval(Duration::fromSeconds(1.0 / 60.0))
 {
     setCentralWidget(main_widget);
+
+    connect(update_timer, &QTimer::timeout, this, &Visualizer::updateVisualizer);
+    update_timer->start(update_timer_interval.getMilliseconds());
 }
 
-Visualizer::~Visualizer()
```
Comment (jonathanlew) [inline src/software/visualizer/widgets/visualizer.cpp:8]: Why don't we need this destructor anymore?

File: src/software/visualizer/widgets/main_widget.cpp [hunk 3]
```diff
@@ -39,25 +39,19 @@ MainWidget::MainWidget(QWidget* parent)
     update();
 }
 
-MainWidget::~MainWidget()
```
Comment (jonathanlew) [inline src/software/visualizer/widgets/main_widget.cpp:42]: Why don't we need this destructor?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looking good just got a couple questions

File: src/software/visualizer/widgets/visualizer.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/visualizer/widgets/visualizer.cpp]: fixed

File: src/software/visualizer/visualizer_wrapper.cpp [hunk 4]
```diff
@@ -1,115 +1,104 @@
...
       ThreadedObserver<PlayInfo>(),
       ThreadedObserver<RobotStatus>(),
-      termination_promise_ptr(std::make_shared<std::promise<void>>())
+      termination_promise_ptr(std::make_shared<std::promise<void>>()),
+      // We want to show the most recent world and AI data, but also want things to look
+      // smooth if the stream of data isn't perfectly consistent, so we use a very small
+      // buffer of 2 values to be responsive while also giving a small buffer for
+      // smoothness
+      world_draw_functions_buffer(
+          std::make_shared<ThreadSafeBuffer<WorldDrawFunction>>(2)),
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/visualizer/visualizer_wrapper.cpp]: fixed

File: src/software/visualizer/visualizer_wrapper.h [hunk 5]
```diff
@@ -47,56 +49,33 @@ class VisualizerWrapper : public ThreadedObserver<World>,
...
-    WorldDrawFunction most_recent_world_draw_function;
-    AIDrawFunction most_recent_ai_draw_function;
-
     std::thread run_visualizer_thread;
     std::shared_ptr<std::promise<void>> termination_promise_ptr;
 
-    std::shared_ptr<Visualizer> visualizer;
-    std::shared_ptr<QApplication> application;
+    // Buffers that are shared with the instance of the Visualizer so that data can
+    // be passed safely
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/visualizer/visualizer_wrapper.h:72]: Full jdoc comments or line comments? What needs clarification here?

File: src/software/visualizer/widgets/zoomable_qgraphics_view.h [hunk 6]
```diff
@@ -17,6 +17,5 @@ class ZoomableQGraphicsView : public QGraphicsView
     void wheelEvent(QWheelEvent *event);
 
    private:
-    const double zoom_in_scaling_factor  = 1.04;
-    const double zoom_out_scaling_factor = 0.96;
+    const double scaling_factor_base = 1.0005;
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/visualizer/widgets/zoomable_qgraphics_view.h]: tried to make it more clear, lmk if that's better.

File: src/software/visualizer/widgets/visualizer.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/visualizer/widgets/visualizer.cpp:8]: We actually don't need to manually destroy the raw pointers (Qt takes ownership), so have no need for the destructor

File: src/software/visualizer/widgets/main_widget.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/visualizer/widgets/main_widget.cpp:42]: same as above

**Review by henryymliu — DISMISSED** [review]
LGTM

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by jsn0716 — DISMISSED** [review]
LGTM!

**Comment by MathewMacDougall** [comment]
Fixed formatting, will need re-approval

**Review by jsn0716 — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**